src/Compilers/Core/Portable/Text/CompositeText.cs

```diff
@@ -373,5 +381,181 @@ private static void TrimInaccessibleText(ArrayBuilder<SourceText> segments)
+                    if (segment[^1] == '\r' &&
+                        i < compositeText.Segments.Length - 1)
+                    {
+                        var nextSegment = compositeText.Segments[i + 1];
+                        Debug.Assert(nextSegment.Length > 0);
+                        if (nextSegment[0] == '\n')
+                        {
+                            accumulatedLineCount -= 1;
+                        }
+                    }
+                }
+
+                _compositeText = compositeText;
+                _segmentLineNumbers = segmentLineNumbers.ToImmutableAndFree();
+
+                // Add one to the accumulatedLineCount for our stored line count (so that we maintain the 
+                // invariant that a text's line count is one greater than the number of newlines it contains)
+                _lineCount = accumulatedLineCount + 1;
+            }
+
+            public override int Count => _lineCount;
+
+            /// <summary>
+            /// Determines the line number of a position in this CompositeText
+            /// </summary>
+            public override int IndexOf(int position)
+            {
+                if (position < 0 || position > _compositeText.Length)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(position));
+                }
+
+                _compositeText.GetIndexAndOffset(position, out var segmentIndex, out var segmentOffset);
+
+                var segment = _compositeText.Segments[segmentIndex];
+                var lineNumberWithinSegment = segment.Lines.IndexOf(segmentOffset);
+
+                return _segmentLineNumbers[segmentIndex] + lineNumberWithinSegment;
+            }
+
+            public override TextLine this[int lineNumber]
+            {
+                get
+                {
+                    if (lineNumber < 0 || lineNumber >= _lineCount)
+                    {
+                        throw new ArgumentOutOfRangeException(nameof(lineNumber));
+                    }
+
+                    // Determine the indices for segments that contribute to our view of the requested line's contents
+                    GetSegmentIndexRangeContainingLine(lineNumber, out var firstSegmentIndexInclusive, out var lastSegmentIndexInclusive);
+
+                    var firstSegmentFirstLineNumber = _segmentLineNumbers[firstSegmentIndexInclusive];
+                    var firstSegment = _compositeText.Segments[firstSegmentIndexInclusive];
+                    var firstSegmentOffset = _compositeText._segmentOffsets[firstSegmentIndexInclusive];
+                    var firstSegmentTextLine = firstSegment.Lines[lineNumber - firstSegmentFirstLineNumber];
+
+                    var lineLength = firstSegmentTextLine.SpanIncludingLineBreak.Length;
+
+                    // walk forward through segments between firstSegmentIndexInclusive and lastSegmentIndexInclusive, and add their
+                    // view of the length of this line.
```

Comment:
might want to mention that this is explicitly *not* getting lastSegmentInclusive, even though we call it inclusive.  since the last segment is handled after this.